src/_pytest/python.py

```diff
@@ -244,7 +244,12 @@ def pytest_pycollect_makeitem(collector, name, obj):
             )
         elif getattr(obj, "__test__", True):
             if is_generator(obj):
-                res = Generator(name, parent=collector)
+                res = Function(name, parent=collector)
+                res.add_marker(
+                    MARK_GEN.xfail(
```

Comment:
is `xfail` enough here?  should these fail discovery?

Reply:
id like to trigger warnings but no failures

Reply:
@nicoddemus does the stdlib warning integration have a mechanism to point to the **location** of a item - if not we should definitively add a helper wrapper for warn_explitict

Reply:
Yep, I'm implementing a `std_warn` function in `Item`. 👍 

This can then be changed to:

```python
res = Function(name, parent=collector)res
res.std_warn(deprecated.YIELD_TESTS.format(name=name), PytestDeprecationWarning)
```

Reply:
Oooh looking at this cleanup I can only dream about the kind of cleanup we will be able to do once we drop Python 2. 😍